packages/directlinespeech/src/patchDialogServiceConnectorInline.js

```diff
@@ -0,0 +1,66 @@
+import createDeferred from 'p-defer';
+
+// Patching a function to add pre-processing of arguments and post-processing of result.
+function patchFunction(fn, pre, post) {
+  return (...args) => {
+    args = pre ? pre(...args) : args;
+
+    const result = fn(...args);
+
+    return post ? post(result) : result;
+  };
+}
+
+export default function patchDialogServiceConnectorInline(dialogServiceConnector) {
```

Comment:
I implemented this, I'd call it `DialogServiceConnectorPatched` that is either a function export or a class export. Why do you like it this way?

Reply:
I think you mean creating something like this:

```js
export default class DialogServiceConnectorPatched extends DialogServiceConnector {
  listenOnceAsync(resolve, reject) {
    super.listenOnceAsync(...); // and we patch their callback
  }
}
```

I am hitting compilation problem due to configuration of our build

```
TypeError: Class constructor DialogServiceConnector cannot be invoked without 'new'
```

Our Babel configuration turn our `DialogServiceConnectorPatched` into a non-class function (i.e. plain old JS mimicking object inheritance thru `prototype`/`constructor`).

But their exported code is ES 2015 (thru `package.json/module` instead of `package.json/main`). So we are importing `DialogServiceConnector` as a *real class*, while our `DialogServiceConnectorPatched` is a *prototype function*.

When our class `DialogServiceConnectorPatched` tries to instantiate its parent `DialogServiceConnector`, it use prototypical way to instantiate it, i.e. not using `new DialogServiceConnector()`, but the plain old way. Thus, it's causing the error.

I am okay to change to your style and it produce lesser and more readable code. But need help on configuring the Babel/Webpack pipeline correctly.

If you are okay, I prefer to file a bug and we fix it later.

Reply:
This is the code that has better readability, leaving it here.

```js
import { DialogServiceConnector } from 'microsoft-cognitiveservices-speech-sdk';
import createDeferred from 'p-defer';

export default class DialogServiceConnectorPatched extends DialogServiceConnector {
  listenOnceAsync(resolve, reject) {
    this._lastRecognitionDeferred = createDeferred();

    super.listenOnceAsync(
      result => {
        this._lastRecognitionDeferred.resolve(result);

        return resolve(result);
      },
      error => {
        this._lastRecognitionDeferred.reject(error);

        return reject(error);
      }
    );
  }

  startContinuousRecognitionAsync(resolve, reject) {
    super.listenOnceAsync(
      () => {},
      error => {
        resolve = null;
        reject && reject(error);
      }
    );

    setTimeout(() => {
      reject = null;
      resolve && resolve();
    }, 0);
  }
}
```